src/transformers/models/llama/modeling_llama.py

```diff
@@ -63,37 +64,185 @@ def _get_unpad_data(padding_mask):
+        causal, a causal mask will be added. If cached 4D attention mask can be reused, no new memory will be
+        allocated.
+        """
+        mask_2d_hash = self._hash_tensor(attention_mask_2d, (query_length, key_value_length))
+
+        # If the attention_mask does not match, but there is still a tensor in the cache, empty the cache
+        if mask_2d_hash not in self.cache_4d_mask and len(self.cache_4d_mask) > 0:
+            self.cache_4d_mask = {}
+
+        # If attention_mask is not cached, create a new one and cache it
+        if mask_2d_hash not in self.cache_4d_mask:
+            input_shape = (attention_mask_2d.shape[0], query_length)
+            past_key_values_length = key_value_length - query_length
+
+            # create causal mask
+            # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+            causal_4d_mask = None
+            if (input_shape[-1] > 1 or self.sliding_window is not None) and self.is_causal:
+                past_key_values_length = key_value_length - query_length
+                causal_4d_mask = self._make_causal_mask(
+                    input_shape,
+                    dtype,
+                    device=attention_mask_2d.device,
+                    past_key_values_length=past_key_values_length,
+                    sliding_window=self.sliding_window,
+                )
+            elif self.sliding_window is not None:
+                raise NotImplementedError("Sliding window is currently only implemented for causal masking")
+
+            # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+            expanded_attn_mask = self._expand_mask(attention_mask_2d, dtype, tgt_len=input_shape[-1]).to(
+                attention_mask_2d.device
+            )
+            cached_4d_mask = expanded_attn_mask if causal_4d_mask is None else expanded_attn_mask + causal_4d_mask
+
+            self.cache_4d_mask[mask_2d_hash] = cached_4d_mask
+
+        return self.cache_4d_mask[mask_2d_hash]
+
+    def _make_causal_mask(
+        self,
+        input_ids_shape: torch.Size,
+        dtype: torch.dtype,
+        device: torch.device,
+        past_key_values_length: int = 0,
+        sliding_window: Optional[int] = None,
+    ):
+        """
+        Make causal mask used for bi-directional self-attention.
+        """
+        bsz, tgt_len = input_ids_shape
+        mask = torch.full((tgt_len, tgt_len), torch.finfo(dtype).min, device=device)
+        mask_cond = torch.arange(mask.size(-1), device=device)
+        mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
+
+        mask = mask.to(dtype)
+
+        if past_key_values_length > 0:
+            mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype, device=device), mask], dim=-1)
+
+        # add lower triangular sliding window mask if necessary
+        if sliding_window is not None:
+            diagonal = past_key_values_length - sliding_window + 1
```

Comment:
@ArthurZucker this corrects a bug that was in Mistral's windowed attention. 

Let's take a look at the following picture:
![Screenshot from 2023-10-19 21-29-39](https://github.com/huggingface/transformers/assets/23423619/52371de8-be8e-46d2-96c2-28b974b03fbe)

Now let's compute the mask as it was done previously with `-sliding_window`

```py
import torch
print(torch.triu(torch.ones(5,5), diagonal=-3))
```
This gives:

```
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [0., 1., 1., 1., 1.]])
```

which is incorrect.

Doing:
```py
import torch
print(torch.triu(torch.ones(5,5), diagonal=-3+1))
```
gives the correct mask:
```
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [0., 1., 1., 1., 1.],
        [0., 0., 1., 1., 1.]])
```

=> Hence we have a +1 now. Also previously the mask didn't work correctly in case the cache + input > sliding window which could happen if a large prompt is added to a system prompt. This is also corrected here by taking `past_key_values_length` into account.

Reply:
Good for me!